README
This is md5deep, a set of cross-platform tools to computer hashes, or
message digests, for any number of files while optionally recursively
digging through the directory structure.  It can also take a list of known
hashes and display the filenames of input files whose hashes either do or
do not match any of the known hashes. This version supports MD5, SHA-1,
SHA-256, Tiger, and Whirlpool hashes. 

See the file NEWS for a list of changes between releases.

See the file COPYING for information about the licensing for this program. 

See the file INSTALL for (generic) compilation and installation
instructions. Here's the short version:

$ ./configure
$ make
$ make install

Note that you must be normally root to install to the default location.
The sudo command is helpful for doing so. You can specify an alternate
installation location using the --prefix option to the configure script.
For example, to install to /home/foo/bin, use:

$ ./configure --prefix=/home/foo

There is complete documentation on how to use the program on the 
project's homepage, http://md5deep.sourceforge.net/ 

== md5deep vs. hashdeep ==

For historical reasons, the program has different options and features
when run with the names "hashdeep" and "md5deep."

hashdeep has a feature called "audit" which:
 * Can also use a list of known hashes to audit a set of FILES. Errors
   are reported to standard error. If no FILES are specified, reads from
   standard input.

 -a Audit mode. Each input file is compared against the set of knowns. An
    audit is said to pass if each input file is matched against exactly
    one file in set of knowns. Any collisions, new files, or missing files
    will make the audit fail. Using this flag alone produces a message,
    either "Audit passed" or "Audit Failed". 

    -v - prints the number of files in each category
    -v -v = prints all discrepancies
    -v -v -v = prints the results for every file examined and every known file. 

 -k <file> - The -k option must be used to load the audit file

To perform an audit:
  hashdeep -r dir  > /tmp/auditfile            # Generate the audit file
  hashdeep -a k /tmp/auditfile -r dir          # test the audit

Notice that the audit is performed with a standard hashdeep output
file. (Internally, the audit is computed as part of the hashing process.)

== Unicode Issues ==
POSIX-based modern computer systems consider filenames to be a
sequence of bytes that are rendered as the application wishes. This
means that filenames typically contain ASCII but can contain UTF-8,
UTF-16, latin1, or even invalid Unicode codings.

Windows-based systems have one set of API calls for ASCII-based
filenames and another set for filenames encoded as UCS-2, which
"produces a fixed-length format by simply using the code point as the
16-bit code unit and produces exactly the same result as UTF-16 for
63,488 code points in the range 0-0xFFFF" according to wikipedia.
(http://en.wikipedia.org/wiki/UTF-16/UCS-2) But wikipedia disputes the
factual accuracy of this statement on the talk page. it's pretty clear
that nobody is entirely sure that Windows actually does, and Windows
itself may not be consistent.

Version 3 of this program addressed this issue by using the TCHAR
variable to hold filenames on Windowa dn by refusing to print them,
priting a "?" instead. Version 4 of this program translates TCHAR
strings to std::string strings at the soonest opportunity using the
Windows function WideCharToMultiByte
(http://msdn.microsoft.com/en-us/library/dd374130%28v=vs.85%29.aspx). Flags
have been added escape Unicode when it is printed.

There is no way (apparently) on Windows to open a UTF-8 filename; it needs to be 
converted back to a multi-byte filename with MultiByteToWideChar.

Fortunately, we never really need to convert back.

Notice that on Windows the files hashed can have unicode characters
but the file with the hashes must have an ASCII name.


COMPILING FOR WINDOWS:
	  -D_UNICODE causes TCHAR to be defined as 'wchar_t'.

COMPILING FOR POSIX:
	  -D_UNICODE is not defined, causing TCHAR to be defined as 'char'.

Previously, win32 functions were controlled with #ifdef statements, like this:

#ifdef _WIN32
  _wfullpath(d_name,fn,PATH_MAX);
#else
  if (NULL == realpath(fn,d_name))
    return TRUE;
#endif

There was also a file called tchar-local.h which actually changed the semantics
of functions on different platforms, with things like this:

   #define  _tcsncpy   strncpy
   #define  _tstat_t   struct stat

This made the code very difficult to maintain.

With the 4.0 rewrite, we have changed this code with C++ functions that return
objects were possible and avoid the use of #defines that so that on _WIN32 systems
the function realpath() gets defined prior to its use, and the mainline code
lacks the realpath() function. You can see this in cycles.cpp:

/* Return the canonicalized absolute pathname in UTF-8 on Windows and POSIX systems */
std::string get_realpath(const TCHAR *fn)
{
#ifdef _WIN32    
    /*
     * expand a relative path to the full path.
     * http://msdn.microsoft.com/en-us/library/506720ff(v=vs.80).aspx
     */
    TCHAR absPath[PATH_MAX];
    if(_fullpath(absPath,fn,PAT_HMAX)==0) return "";
    return tchar_to_utf8(absPath);
#else
    char resolved_name[PATH_MAX];	//
    if(realpath(fn,resolved_name)==0) return "";
    return string(resolved_name);
#endif
}


You can install mingw and then simply configure with something like this:
$ export PATH=$PATH:/usr/local/i386-mingw32-4.3.0/bin
$ ./configure --host=i386-mingw32


== Hash Algorithm References ==

The MD5 algorithm is defined in RFC 1321:
http://www.ietf.org/rfc/rfc1321.txt

The SHA1 algorithm is defined in FIPS 180-1:
http://www.itl.nist.gov/fipspubs/fip180-1.htm

The SHA256 algorithm is defined FIPS 180-2:
http://csrc.nist.gov/publications/fips/fips180-2/fips180-2.pdf

The Tiger algorithm is defined at:
http://www.cs.technion.ac.il/~biham/Reports/Tiger/

The Whirlpool algorithm is defined at:
http://planeta.terra.com.br/informatica/paulobarreto/WhirlpoolPage.html

================================================================
Theory of operation.

main.cpp    - sets up the system.
dig.cpp     - iterates through the individual directories.
 	    - calls hash_file() in hash.cpp for each file to hash
hash.cpp    - performs the hashing of each file.
display.cpp - stores/displays the results.


NEWS
** Changes in version 4.4 (29 Jan 2014)

* New Features

      Added case insensitive auditing with -E mode

* Bug Fixes

      Added checks against user errors on Hashdeep command line

      Fixed bug for known filenames with the directory in the filename

      Fixed bug for double dots in the filename



** Changes in version 4.3 (24 Oct 2012)

* Bug Fixes

      Fixed check for endian-ness, affecting hash generation on
      big-endian platforms.

      Fixed minor bugs related to OpenSolaris.



** Changes in version 4.2 (11 Jun 2012)

* Bug Fixes

      Fixed padding in Tiger hashes for large files



** Changes in version 4.1.1 (5 Jun 2012)

* Bug Fixes

      Fixed hashdeep bug for known files with commas in their names.
      Cleaned up usage message


** Changes in version 4.1.0 (14 Feb 2012)

* New Features

      Added expert mode option to parse Windows PE files

* Bug Fixes

      Fixed junction point handling on Win32



** Changes in version 4.0.1 (22 Jan 2012)

* Bug Fixes

        Fixed hang on DFXML generation on Win32
	Fixed incorrect hashes via stdin on Win32
	Fixed "Too many open files" error on OS X
	Doc files in Win32 have been corrected.



** Changes in version 4.0 (6 Jan 2012)

* New Features

      	Rewrote most of the program in C++.

	Enabled multiprocessor support on all platforms.

	Removed ten character limit on file size mode

* Bug Fixes

	Too numerous to count!



** Changes in version 3.9.1 (31 May 2011)

* Bug Fixes

      Fixed a cosmetic bugs regarding processing standard input
      in piecewise mode.

      Fixed a bug regarding processing known hash files which have
      parenthesis in the filename.

      Clarified documentation regarding multipliers in piecewise
      mode. They are IEC multipliers, or blocks of 1,024. Side note:
      also spelled 'multipliers' correctly.



** Changes in version 3.9 (21 Apr 2011)

* Bug Fixes

      Fixed a critical bug in hashdeep regarding processing
      standard input



** Changes in version 3.8 (15 Apr 2011)

* New Features

      Added -f mode to get the list of input files from
      a file on the disk.

      Added 64-bit executables for Windows.


* Bug Fixes

      Fixed a cosmetic bug regarding the error messages for
      piecewise mode.




** Changes in version 3.7 (18 Dec 2010)

* New Features

      Added triage mode, -Z flag, to compute file size and partial
      hashes in addition to full file hash.

      Updated Windows cross-compiler to MinGW-4.2.1


* Bug Fixes

      Fixed hashdeep piecewise mode to include the correct size
      of each piece being hashed. This enables audit mode to
      work correctly.

      Addressed display problem for time estimates of files with
      long filenames in the current directory.

      Added "Input/Output error" to the list of fatal errors when
      processing files




** Changes in version 3.6 (23 Mar 2010)

* New Features

      Added expert mode found in md5deep to hashdeep.

* Bug Fixes

      Fixed error checking in helper routines.




** Changes in version 3.5.1 (30 Dec 2009)

* Bug Fixes

      Fixed processing of zero byte files.




** Changes in version 3.5 (22 Dec 2009)

* New Features

      Added mode to load list of files to process from a file (-f mode)

* Bug Fixes

      Fixed file size computation for piecewise hashing of standard input.

      Fixed error when auditing a directory to check for identical copies
      of a file if one copy is removed.



** Changes in version 3.4 (11 Jun 2009)

* Bug Fixes

      Fixed directory handle leak during recursive parsing on Windows

      Fixed -n mode (display unused hashes)



** Changes in version 3.3 (4 Apr 2009)

* New Features

* Bug Fixes

      Skip all reparse points (symbolic links and junction points)
      on Windows. Required for sane parsing of Windows Vista or 7 systems.

      Fixed handling of 64-bit timestamps on Windows.




** Changes in version 3.2 (25 Jan 2009)

* New Features

	Added Comma Separated Value (CSV) mode.


* Bug Fixes

	Fixed support for 64-bit files on Windows

	Fixed hashdeep input file parsing to allow filenames with
	commas and properly handle other kinds of errors. Also
	modified the invocation listing in output files such that
        even long lines can be successfully parsed.




** Changes in version 3.1 (23 Jul 2008)

* New Features

	Added size threshold modes to only process files smaller
	than a given size.

	Added a timestamp mode that records the creation time time
	for each file on Win32, the change time on all other operating
	systems.

	Streamlined file size computation process, which makes
	the programs about 15% faster.


* Bug Fixes

	Corrected time estimates for large files (e.g. files which
        require more than one day).




** Changes in version 3.0 (10 May 2008)

* New Features

	Added hashdeep program to support multihashing and
        hash file auditing

	Added support for new iLook style hashes

* Bug Fixes

	Fixed obscure bug that caused a crash (double free) when
	attempting to check a very small file for EnCase hashes



** Changes in version 2.0.1 (31 Oct 2007)

* Bug Fixes

	Added support for older Linux systems that don't have the
	latest and greatest ioctl functions.



** Changes in version 2.0 (15 Oct 2007)

* New Features

	Using GNU Autotools for configuration and compilation. This
	should help avoid platform specific issues such as SHA-1
	problems previously found on 64-bit versions of AIX.

	Added support for files with Unicode characters in their
	filenames on Microsoft Windows.

	Added support for EnCase hash sets (.hash files).

	Updated web site and quick start guide

	Slightly reduced the size of all of the executables by removing
	duplicated code.

* Bug Fixes

	Fixed time estimation mode for block devices on OS X and Linux

	Fixed cosmetic error where estimated time remaining mode is
	being used in conjunction with piecewise hashing. Time estimates

sha1.c

/* MD5DEEP - sha1.c
 *
 * By Jesse Kornblum
 *
 * This is a work of the US Government. In accordance with 17 USC 105,
 * copyright protection is not available for any work of the US Government.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 */

/* $Id: sha1.c,v 1.6 2007/09/23 01:54:26 jessekornblum Exp $ */

#include "main.h"
#include "sha1.h"

int setup_hashing_algorithm(state *s)
{
  s->hash_length        = 20;
  s->hash_init          = hash_init_sha1;
  s->hash_update        = hash_update_sha1;
  s->hash_finalize      = hash_final_sha1;
  
  s->h_plain = s->h_bsd = s->h_md5deep_size = 1;      
  s->h_ilook = s->h_hashkeeper = 0;
  s->h_nsrl15 = 1;
  s->h_nsrl20 = 1;
  s->h_encase = 0;
  
  s->hash_context = (context_sha1_t *)malloc(sizeof(context_sha1_t));
  if (NULL == s->hash_context)
    return TRUE;
  
  return FALSE;
}



/* This implementation of SHA1-1 was written by 
   Steve Reid <steve@edmweb.com> and was included in this program on 
   the good faith that it is public domain */

/* Test Vectors (from FIPS PUB 180-1)
"abc"
  A9993E36 4706816A BA3E2571 7850C26C 9CD0D89D
"abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
  84983E44 1C3BD26E BAAE4AA1 F95129E5 E54670F1
A million repetitions of "a"
  34AA973C D4C4DAA4 F61EEB2B DBAD2731 6534016F
*/


#define rol(value, bits) (((value) << (bits)) | ((value) >> (32 - (bits))))

/* blk0() and blk() perform the initial expand. */
/* I got the idea of expanding during the round function from SSLeay */
#ifndef WORDS_BIGENDIAN
#define blk0(i) (block->l[i] = (rol(block->l[i],24)&0xFF00FF00) \
    |(rol(block->l[i],8)&0x00FF00FF))
#else
#define blk0(i) block->l[i]
#endif
#define blk(i) (block->l[i&15] = rol(block->l[(i+13)&15]^block->l[(i+8)&15] \
    ^block->l[(i+2)&15]^block->l[i&15],1))

/* (R0+R1), R2, R3, R4 are the different operations used in SHA1 */
#define R0(v,w,x,y,z,i) z+=((w&(x^y))^y)+blk0(i)+0x5A827999+rol(v,5);w=rol(w,30);
#define R1(v,w,x,y,z,i) z+=((w&(x^y))^y)+blk(i)+0x5A827999+rol(v,5);w=rol(w,30);
#define R2(v,w,x,y,z,i) z+=(w^x^y)+blk(i)+0x6ED9EBA1+rol(v,5);w=rol(w,30);
#define R3(v,w,x,y,z,i) z+=(((w|x)&y)|(w&x))+blk(i)+0x8F1BBCDC+rol(v,5);w=rol(w,30);
#define R4(v,w,x,y,z,i) z+=(w^x^y)+blk(i)+0xCA62C1D6+rol(v,5);w=rol(w,30);


int hash_init_sha1(state *s)
{
  SHA1Init(s->hash_context);
  return FALSE;
}

int hash_update_sha1(state *s, unsigned char *buf, uint64_t len)
{
  SHA1Update(s->hash_context, buf, len);
  return FALSE;
}

int hash_final_sha1(state *s, unsigned char *digest)
{
  SHA1Final(digest,s->hash_context);
  return FALSE;
}



/* Hash a single 512-bit block. This is the core of the algorithm. */

void SHA1Transform(uint32_t state[5], unsigned char buffer[64])
{
  uint32_t a, b, c, d, e;
  typedef union {
    unsigned char c[64];
    uint32_t l[16];
  } CHAR64LONG16;
  CHAR64LONG16* block;
#ifdef SHA1HANDSOFF
  static unsigned char workspace[64];
  block = (CHAR64LONG16*)workspace;
  memcpy(block, buffer, 64);
#else
  block = (CHAR64LONG16*)buffer;
#endif
  /* Copy context->state[] to working vars */
  a = state[0];
  b = state[1];
  c = state[2];
  d = state[3];
  e = state[4];
  /* 4 rounds of 20 operations each. Loop unrolled. */
  R0(a,b,c,d,e, 0); R0(e,a,b,c,d, 1); R0(d,e,a,b,c, 2); R0(c,d,e,a,b, 3);
  R0(b,c,d,e,a, 4); R0(a,b,c,d,e, 5); R0(e,a,b,c,d, 6); R0(d,e,a,b,c, 7);
  R0(c,d,e,a,b, 8); R0(b,c,d,e,a, 9); R0(a,b,c,d,e,10); R0(e,a,b,c,d,11);
  R0(d,e,a,b,c,12); R0(c,d,e,a,b,13); R0(b,c,d,e,a,14); R0(a,b,c,d,e,15);
  R1(e,a,b,c,d,16); R1(d,e,a,b,c,17); R1(c,d,e,a,b,18); R1(b,c,d,e,a,19);
  R2(a,b,c,d,e,20); R2(e,a,b,c,d,21); R2(d,e,a,b,c,22); R2(c,d,e,a,b,23);
  R2(b,c,d,e,a,24); R2(a,b,c,d,e,25); R2(e,a,b,c,d,26); R2(d,e,a,b,c,27);
  R2(c,d,e,a,b,28); R2(b,c,d,e,a,29); R2(a,b,c,d,e,30); R2(e,a,b,c,d,31);
  R2(d,e,a,b,c,32); R2(c,d,e,a,b,33); R2(b,c,d,e,a,34); R2(a,b,c,d,e,35);
  R2(e,a,b,c,d,36); R2(d,e,a,b,c,37); R2(c,d,e,a,b,38); R2(b,c,d,e,a,39);
  R3(a,b,c,d,e,40); R3(e,a,b,c,d,41); R3(d,e,a,b,c,42); R3(c,d,e,a,b,43);
  R3(b,c,d,e,a,44); R3(a,b,c,d,e,45); R3(e,a,b,c,d,46); R3(d,e,a,b,c,47);
  R3(c,d,e,a,b,48); R3(b,c,d,e,a,49); R3(a,b,c,d,e,50); R3(e,a,b,c,d,51);
  R3(d,e,a,b,c,52); R3(c,d,e,a,b,53); R3(b,c,d,e,a,54); R3(a,b,c,d,e,55);
  R3(e,a,b,c,d,56); R3(d,e,a,b,c,57); R3(c,d,e,a,b,58); R3(b,c,d,e,a,59);
  R4(a,b,c,d,e,60); R4(e,a,b,c,d,61); R4(d,e,a,b,c,62); R4(c,d,e,a,b,63);
  R4(b,c,d,e,a,64); R4(a,b,c,d,e,65); R4(e,a,b,c,d,66); R4(d,e,a,b,c,67);
  R4(c,d,e,a,b,68); R4(b,c,d,e,a,69); R4(a,b,c,d,e,70); R4(e,a,b,c,d,71);
  R4(d,e,a,b,c,72); R4(c,d,e,a,b,73); R4(b,c,d,e,a,74); R4(a,b,c,d,e,75);
  R4(e,a,b,c,d,76); R4(d,e,a,b,c,77); R4(c,d,e,a,b,78); R4(b,c,d,e,a,79);
  /* Add the working vars back into context.state[] */
  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;
  state[4] += e;
  /* Wipe variables */
  a = b = c = d = e = 0;
}


/* SHA1Init - Initialize new context */

void SHA1Init(SHA1_CTX* context)
{
  /* SHA1 initialization constants */
  context->state[0] = 0x67452301;
  context->state[1] = 0xEFCDAB89;
  context->state[2] = 0x98BADCFE;
  context->state[3] = 0x10325476;
  context->state[4] = 0xC3D2E1F0;
  context->count[0] = context->count[1] = 0;
}


/* Run your data through this. */
void SHA1Update(SHA1_CTX* context, unsigned char * data, unsigned int len)
{
  unsigned int i, j;

  j = (context->count[0] >> 3) & 63;

  if ((context->count[0] += len << 3) < (len << 3)) 
    context->count[1]++;

  context->count[1] += (len >> 29);

  if ((j + len) > 63) 
  {
    memcpy(&context->buffer[j], data, (i = 64-j));
    SHA1Transform(context->state, context->buffer);
    for ( ; i + 63 < len; i += 64) {
      SHA1Transform(context->state, &data[i]);
    }
    j = 0;
  }
  else i = 0;
  memcpy(&context->buffer[j], &data[i], len - i);
}


/* Add padding and return the message digest. */

void SHA1Final(unsigned char digest[20], SHA1_CTX* context)
{
  uint32_t i, j;
  unsigned char finalcount[8];

  for (i = 0; i < 8; i++) {
    finalcount[i] = (unsigned char)((context->count[(i >= 4 ? 0 : 1)]
				     >> ((3-(i & 3)) * 8) ) & 255);  /* Endian independent */
  }
  SHA1Update(context, (unsigned char *)"\200", 1);
  while ((context->count[0] & 504) != 448) {
    SHA1Update(context, (unsigned char *)"\0", 1);
  }
  SHA1Update(context, finalcount, 8);  /* Should cause a SHA1Transform() */
  for (i = 0; i < 20; i++) {
    digest[i] = (unsigned char)
      ((context->state[i>>2] >> ((3-(i & 3)) * 8) ) & 255);
  }
  /* Wipe variables */
  i = j = 0;
  memset(context->buffer, 0, 64);
  memset(context->state, 0, 20);
  memset(context->count, 0, 8);
  memset(&finalcount, 0, 8);
#ifdef SHA1HANDSOFF  /* make SHA1Transform overwrite it's own static vars */
  SHA1Transform(context->state, context->buffer);
#endif
}


